@m-blaha I got the following errors when I built it from source, ran dbus and the built daemon in a container and rebuilt libdnf with the daemon tests enabled:
```
======================================================================
FAIL: test_list_repos (test_repolist.RepoTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/dbox/libdnf/test/dnfdaemon-server/test_repolist.py", line 52, in test_list_repos
    self.assertEqual(
AssertionError: dbus.[32 chars]ing('enabled'): dbus.Boolean(True, variant_lev[441 chars]v}')) != dbus.[32 chars]ing('name'): dbus.String('Repository 1', varia[433 chars]v}'))

======================================================================
FAIL: test_list_repos_spec (test_repolist.RepoTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/dbox/libdnf/test/dnfdaemon-server/test_repolist.py", line 72, in test_list_repos_spec
    self.assertEqual(
AssertionError: dbus.Array([], signature=dbus.Signature('a{sv}')) != dbus.Array([dbus.Dictionary({dbus.String('name'): dbu[226 chars]v}'))

======================================================================
FAIL: test_repoquery_all (test_repoquery.RepoTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/dbox/libdnf/test/dnfdaemon-server/test_repoquery.py", line 59, in test_repoquery_all
    self.assertCountEqual(
AssertionError: Element counts were not equal:

Diff is 1723 characters long. Set self.maxDiff to None to see it.

======================================================================
FAIL: test_repoquery_match (test_repoquery.RepoTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/dbox/libdnf/test/dnfdaemon-server/test_repoquery.py", line 92, in test_repoquery_match
    self.assertCountEqual(
AssertionError: Element counts were not equal:

Diff is 861 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 8 tests in 8.048s

FAILED (failures=4)
```

--- a/dnfdaemon-client/commands/repolist/repolist.cpp
+++ b/dnfdaemon-client/commands/repolist/repolist.cpp
@@ -26,6 +26,7 @@ along with dnfdaemon-client.  If not, see <https://www.gnu.org/licenses/>.
 #include <libsmartcols/libsmartcols.h>
 
 #include <iostream>
+#include <numeric>
 
 namespace dnfdaemon::client {
 
@@ -90,13 +91,59 @@ void CmdRepolist::set_argument_parser(Context & ctx) {
     ctx.arg_parser.get_root_command()->register_command(repolist);
 }
 
+/// Joins vector of strings to a single string using given separator
+/// ["a", "b", "c"] -> "a b c"
+std::string join(const std::vector<std::string> & str_list, const std::string & separator) {
+    return std::accumulate(
+        str_list.begin(), str_list.end(), std::string(),
+        [&](const std::string & a, const std::string & b) -> std::string {
+            return a + (a.length() > 0 ? separator : "") + b;
+        });
+}
+
+/// Joins vector of string pairs to a single string. Pairs are joined using
+/// field_separator, records using record_separator
+/// [("a", "1"), ("b", "2")] -> "a: 1, b: 2"
+std::string join_pairs(const std::vector<std::pair<std::string, std::string>> & pair_list, const std::string & field_separator, const std::string & record_separator) {
+    std::vector<std::string> records {};
+    for (auto & pair: pair_list) {
+        records.emplace_back(pair.first + field_separator + pair.second);
+    }
+    return join(records, record_separator);
+}
+
 class RepoDbus {
 public:
     RepoDbus(dnfdaemon::KeyValueMap & rawdata) : rawdata(rawdata){};
     std::string get_id() const { return rawdata.at("id"); }
     std::string get_name() const { return rawdata.at("name"); }
     bool is_enabled() const { return rawdata.at("enabled"); }
-    uint64_t get_size() const { return rawdata.at("repo_size"); }
+    uint64_t get_size() const { return rawdata.at("size"); }
+    std::string get_revision() const { return rawdata.at("revision"); }
+    std::vector<std::string> get_content_tags() const { return rawdata.at("content_tags"); }
+    std::vector<std::pair<std::string, std::string>> get_distro_tags() const {
+        // sdbus::Variant cannot handle vector<pair<string,string>> so values are
+        // serialized to vector<string>.
+        // convert [tag1, val1, tag2, val2,...] back to [(tag1, val1), (tag2, val2),...]
+        std::vector<std::pair<std::string, std::string>> dt {};
+        std::vector<std::string> tags_raw = rawdata.at("distro_tags");
+        if (!tags_raw.empty()) {
+            for(size_t i = 0; i < (tags_raw.size() - 1); i+=2){
+                dt.emplace_back(tags_raw[i], tags_raw[i + 1]);
+            }
+        }
+        return dt;
+    }
+    int get_max_timestamp() const { return rawdata.at("updated"); }
+    uint64_t get_pkgs() const { return rawdata.at("pkgs"); }
+    uint64_t get_available_pkgs() const { return rawdata.at("available_pkgs"); }
+    std::string get_metalink() const { return rawdata.at("metalink"); }
+    std::string get_mirrorlist() const { return rawdata.at("mirrorlist"); }
+    std::vector<std::string> get_baseurl() const { return rawdata.at("baseurl"); }
+    int get_metadata_expire() const { return rawdata.at("metadata_expire"); }
+    std::vector<std::string> get_excludepkgs() const { return rawdata.at("excludepkgs"); }
+    std::vector<std::string> get_includepkgs() const { return rawdata.at("includepkgs"); }
+    std::string get_repofile() const { return rawdata.at("repofile"); }
 
 private:
     dnfdaemon::KeyValueMap rawdata;
@@ -138,6 +185,7 @@ void CmdRepolist::run(Context & ctx) {
     options["enable_disable"] = enable_disable_option->get_value();
     std::vector<std::string> patterns;
     if (!patterns_options->empty()) {
+        options["enable_disable"] = "all";
         patterns.reserve(patterns_options->size());
         for (auto & pattern : *patterns_options) {
             auto option = dynamic_cast<libdnf::OptionString *>(pattern.get());
@@ -146,14 +194,15 @@ void CmdRepolist::run(Context & ctx) {
     }
     options["patterns"] = patterns;
 
-    options["command"] = command;
-
+    std::vector<std::string> attrs {"id", "name", "enabled"};
     if (command == "repoinfo") {
-        // load repositories
-        if (ctx.wait_for_repos() == dnfdaemon::RepoStatus::ERROR) {
-            throw std::runtime_error("Failed to load repositories");
-        };
+        std::vector<std::string> repoinfo_attrs {
+            "size", "revision", "content_tags", "distro_tags", "updated",
+            "pkgs", "available_pkgs", "metalink", "mirrorlist", "baseurl",
+            "metadata_expire", "excludepkgs", "includepkgs", "repofile"};
+        attrs.insert(attrs.end(), repoinfo_attrs.begin(), repoinfo_attrs.end());
     }
+    options["repo_attrs"] = attrs;
 
     // call list() method on repo interface via dbus
     dnfdaemon::KeyValueMapList repositories;
@@ -181,8 +230,23 @@ void CmdRepolist::run(Context & ctx) {
         // TODO(mblaha): output using smartcols
         for (auto & raw_repo : repositories) {
             RepoDbus repo(raw_repo);
-            std::cout << "REPO: " << repo.get_id() << std::endl;
-            std::cout << "size: " << repo.get_size() << std::endl;
+            std::cout << "Repo-id: " << repo.get_id() << std::endl;
+            std::cout << "Repo-name: " << repo.get_name() << std::endl;
+            std::cout << "Repo-status: " << repo.is_enabled() << std::endl;
+            std::cout << "Repo-revision: " << repo.get_revision() << std::endl;
+            std::cout << "Repo-tags: " << join(repo.get_content_tags(), ", ") << std::endl;
+            std::cout << "Repo-distro-tags: " << join_pairs(repo.get_distro_tags(), ": ", "\n") << std::endl;
+            std::cout << "Repo-updated: " << repo.get_max_timestamp() << std::endl;
+            std::cout << "Repo-pkgs: " << repo.get_pkgs() << std::endl;
+            std::cout << "Repo-available-pkgs: " << repo.get_available_pkgs() << std::endl;
+            std::cout << "Repo-size: " << repo.get_size() << std::endl;
+            std::cout << "Repo-metalink: " << repo.get_metalink() << std::endl;
+            std::cout << "Repo-mirrors: " << repo.get_mirrorlist() << std::endl;
+            std::cout << "Repo-baseurl: " << join(repo.get_baseurl(), "\n") << std::endl;
+            std::cout << "Repo-expire: " << repo.get_metadata_expire() << std::endl;
+            std::cout << "Repo-exclude: " << join(repo.get_excludepkgs(), ", ") << std::endl;
+            std::cout << "Repo-include: " << join(repo.get_includepkgs(), ", ") << std::endl;
+            std::cout << "Repo-filename: " << repo.get_repofile() << std::endl;
         }
     }
 }

--- a/dnfdaemon-server/services/repo/repo.cpp
+++ b/dnfdaemon-server/services/repo/repo.cpp
@@ -32,6 +32,164 @@ along with dnfdaemon-server.  If not, see <https://www.gnu.org/licenses/>.
 #include <iostream>
 #include <string>
 
+// repository attributes available to be retrieved
+// based on `dnf repolist` command
+enum class RepoAttribute {
+    // from repo configuration
+    id,
+    name,
+    enabled,
+    baseurl,
+    metalink,
+    mirrorlist,
+    metadata_expire,
+    excludepkgs,
+    includepkgs,
+    repofile,
+
+    // require metadata loading
+    revision,
+    content_tags,
+    distro_tags,
+    updated,
+    pkgs,
+    available_pkgs, // number of not excluded packages
+    size
+};
+
+std::vector<RepoAttribute> metadata_required_attributes {
+    RepoAttribute::revision,
+    RepoAttribute::content_tags,
+    RepoAttribute::distro_tags,
+    RepoAttribute::updated,
+    RepoAttribute::pkgs,
+    RepoAttribute::available_pkgs,
+    RepoAttribute::size
+};
+
+// map string package attribute name to actual attribute
+const static std::map<std::string, RepoAttribute> repo_attributes {
+    {"id", RepoAttribute::id},
+    {"name", RepoAttribute::name},
+    {"enabled", RepoAttribute::enabled},
+    {"baseurl", RepoAttribute::baseurl},
+    {"metalink", RepoAttribute::metalink},
+    {"mirrorlist", RepoAttribute::mirrorlist},
+    {"metadata_expire", RepoAttribute::metadata_expire},
+    {"excludepkgs", RepoAttribute::excludepkgs},
+    {"includepkgs", RepoAttribute::includepkgs},
+    {"repofile", RepoAttribute::repofile},
+    {"revision", RepoAttribute::revision},
+    {"content_tags", RepoAttribute::content_tags},
+    {"distro_tags", RepoAttribute::distro_tags},
+    {"updated", RepoAttribute::updated},
+    {"pkgs", RepoAttribute::pkgs},
+    {"available_pkgs", RepoAttribute::available_pkgs},
+    {"size", RepoAttribute::size}
+};
+
+// converts Repo object to dbus map
+dnfdaemon::KeyValueMap repo_to_map(libdnf::Base & base, const libdnf::WeakPtr<libdnf::rpm::Repo, false> libdnf_repo, std::vector<std::string> & attributes) {
+    dnfdaemon::KeyValueMap dbus_repo;
+    // attributes required by client
+    for (auto & attr : attributes) {
+        switch (repo_attributes.at(attr)) {
+            case RepoAttribute::id:
+                dbus_repo.emplace(attr, libdnf_repo->get_id());
+                break;
+            case RepoAttribute::name:
+                dbus_repo.emplace(attr, libdnf_repo->get_config().name().get_value());
+                break;
+            case RepoAttribute::enabled:
+                dbus_repo.emplace(attr, libdnf_repo->is_enabled());
+                break;
+            case RepoAttribute::size:
+                {
+                    auto & solv_sack = base.get_rpm_solv_sack();
+                    uint64_t size = 0;
+                    libdnf::rpm::SolvQuery query(&solv_sack);
+                    std::vector<std::string> reponames = {libdnf_repo->get_id()};
+                    query.ifilter_repoid(libdnf::sack::QueryCmp::EQ, reponames);
+                    for (auto pkg : query) {
+                        size += pkg.get_download_size();
+                    }
+                    dbus_repo.emplace(attr, size);
+                }
+                break;
+            case RepoAttribute::revision:
+                dbus_repo.emplace(attr, libdnf_repo->get_revision());
+                break;
+            case RepoAttribute::content_tags:
+                dbus_repo.emplace(attr, libdnf_repo->get_content_tags());
+                break;
+            case RepoAttribute::distro_tags:
+                {
+                    // sdbus::Variant cannot accomodate a std::pair
+                    std::vector<std::string> distro_tags {};
+                    for (auto & dt: libdnf_repo->get_distro_tags()) {
+                        distro_tags.emplace_back(dt.first);
+                        distro_tags.emplace_back(dt.second);
+                    }
+                    dbus_repo.emplace(attr, distro_tags);
+                }
+                break;
+            case RepoAttribute::updated:
+                dbus_repo.emplace(attr, libdnf_repo->get_max_timestamp());
+                break;
+            case RepoAttribute::pkgs:
+                {
+                    auto & solv_sack = base.get_rpm_solv_sack();
+                    libdnf::rpm::SolvQuery query(&solv_sack, libdnf::rpm::SolvQuery::InitFlags::IGNORE_EXCLUDES);
+                    std::vector<std::string> reponames = {libdnf_repo->get_id()};
+                    query.ifilter_repoid(libdnf::sack::QueryCmp::EQ, reponames);
+                    dbus_repo.emplace(attr, query.size());
+                }
+                break;
+            case RepoAttribute::available_pkgs:
+                {
+                    auto & solv_sack = base.get_rpm_solv_sack();
+                    libdnf::rpm::SolvQuery query(&solv_sack);
+                    std::vector<std::string> reponames = {libdnf_repo->get_id()};
+                    query.ifilter_repoid(libdnf::sack::QueryCmp::EQ, reponames);
+                    dbus_repo.emplace(attr, query.size());
+                }
+                break;
+            case RepoAttribute::metalink:
+                {
+                    auto opt = libdnf_repo->get_config().metalink();
+                    dbus_repo.emplace(attr, (opt.empty() || opt.get_value().empty()) ? "" : opt.get_value());
+                }
+                break;
+            case RepoAttribute::mirrorlist:
+                {
+                    auto opt = libdnf_repo->get_config().mirrorlist();
+                    dbus_repo.emplace(attr, (opt.empty() || opt.get_value().empty()) ? "" : opt.get_value());
+                }
+                break;
+            case RepoAttribute::baseurl:
+                dbus_repo.emplace(attr, libdnf_repo->get_config().baseurl().get_value());
+                break;
+            case RepoAttribute::metadata_expire:
+                dbus_repo.emplace(attr, libdnf_repo->get_config().metadata_expire().get_value());
+                break;
+            case RepoAttribute::excludepkgs:
+                dbus_repo.emplace(attr, libdnf_repo->get_config().excludepkgs().get_value());
+                break;
+            case RepoAttribute::includepkgs:
+                dbus_repo.emplace(attr, libdnf_repo->get_config().includepkgs().get_value());
+                break;
+            case RepoAttribute::repofile:
+                dbus_repo.emplace(attr, libdnf_repo->get_repo_file_path());
+                break;
+        }
+    }
+    return dbus_repo;
+}
+
+bool keyval_repo_compare(const dnfdaemon::KeyValueMap & first, const dnfdaemon::KeyValueMap & second) {
+    return key_value_map_get<std::string>(first, "id") < key_value_map_get<std::string>(second, "id");
+}
+
 void Repo::dbus_register() {
     auto dbus_object = session.get_dbus_object();
     dbus_object->registerMethod(
@@ -40,39 +198,36 @@ void Repo::dbus_register() {
         });
 }
 
-uint64_t repo_size(libdnf::rpm::SolvSack & sack, const libdnf::WeakPtr<libdnf::rpm::Repo, false> & repo) {
-    uint64_t size = 0;
-    libdnf::rpm::SolvQuery query(&sack);
-    std::vector<std::string> reponames = {repo->get_id()};
-    query.ifilter_repoid(libdnf::sack::QueryCmp::EQ, reponames);
-    for (auto pkg : query) {
-        size += pkg.get_download_size();
-    }
-    return size;
-}
-
-bool keyval_repo_compare(const dnfdaemon::KeyValueMap & first, const dnfdaemon::KeyValueMap & second) {
-    return key_value_map_get<std::string>(first, "id") < key_value_map_get<std::string>(second, "id");
-}
-
 void Repo::list(sdbus::MethodCall call) {
     dnfdaemon::KeyValueMap options;
     call >> options;
     auto worker = std::thread([this, options = std::move(options), call = std::move(call)]() {
         try {
+            const std::vector<std::string> empty_list {};
+
             // read options from dbus call
             std::string enable_disable = key_value_map_get<std::string>(options, "enable_disable", "enabled");
-            std::vector<std::string> default_patterns{};
             std::vector<std::string> patterns =
-                key_value_map_get<std::vector<std::string>>(options, "patterns", std::move(default_patterns));
-            std::string command = key_value_map_get<std::string>(options, "command", "repolist");
-
-            // repoinfo command needs repositories loaded into sack
-            // TODO(mblaha): check that repos are loaded
+                key_value_map_get<std::vector<std::string>>(options, "patterns", empty_list);
+            // check demanded attributes
+            std::vector<std::string> repo_attrs = key_value_map_get<std::vector<std::string>>(options, "repo_attrs", empty_list);
+            bool fill_sack_needed = false;
+            for (auto & attr_str : repo_attrs) {
+                if (repo_attributes.count(attr_str) == 0) {
+                    throw std::runtime_error(fmt::format("Repo attribute '{}' not supported", attr_str));
+                }
+                if (!fill_sack_needed) {
+                    fill_sack_needed = std::find(metadata_required_attributes.begin(), metadata_required_attributes.end(), repo_attributes.at(attr_str)) != metadata_required_attributes.end();
+                }
+            }
+            // always return repoid
+            repo_attrs.push_back("id");
+            if (fill_sack_needed) {
+                session.fill_sack();
+            }
 
             // prepare repository query filtered by options
             auto base = session.get_base();
-            auto & solv_sack = base->get_rpm_solv_sack();
             auto & rpm_repo_sack = base->get_rpm_repo_sack();
             auto repos_query = rpm_repo_sack.new_query();
 
@@ -90,16 +245,9 @@ void Repo::list(sdbus::MethodCall call) {
 
             // create reply from the query
             dnfdaemon::KeyValueMapList out_repositories;
+
             for (auto & repo : repos_query.get_data()) {
-                dnfdaemon::KeyValueMap out_repo;
-                out_repo.emplace(std::make_pair("id", repo->get_id()));
-                out_repo.emplace(std::make_pair("name", repo->get_config().name().get_value()));
-                out_repo.emplace(std::make_pair("enabled", repo->is_enabled()));
-                // TODO(mblaha): add all other repository attributes
-                if (command == "repoinfo") {
-                    out_repo.emplace(std::make_pair("repo_size", repo_size(solv_sack, repo)));
-                }
-                out_repositories.push_back(std::move(out_repo));
+                out_repositories.push_back(repo_to_map(*base, repo, repo_attrs));
             }
 
             std::sort(out_repositories.begin(), out_repositories.end(), keyval_repo_compare);

--- a/test/dnfdaemon-server/test_repolist.py
+++ b/test/dnfdaemon-server/test_repolist.py
@@ -49,17 +49,17 @@ class RepoTest(unittest.TestCase):
     def test_list_repos(self):
         # get list of all repositories
         self.assertEqual(
-            self.iface_repo.list({}),
+            self.iface_repo.list({"repo_attrs": ["name", "enabled"]}),
             dbus.Array([
                 dbus.Dictionary({
                     dbus.String('name'): dbus.String('Repository 1', variant_level=1),
                     dbus.String('enabled'): dbus.Boolean(True, variant_level=1),
-                    dbus.String('id'): dbus.String('repo1', variant_level=1)},
+                    dbus.String('id'): dbus.String('rpm-repo1', variant_level=1)},
                     signature=dbus.Signature('sv')),
                 dbus.Dictionary({
                     dbus.String('name'): dbus.String('Repository 2', variant_level=1),
                     dbus.String('enabled'): dbus.Boolean(True, variant_level=1),
-                    dbus.String('id'): dbus.String('repo2', variant_level=1)},
+                    dbus.String('id'): dbus.String('rpm-repo2', variant_level=1)},
                     signature=dbus.Signature('sv')),
                 ],
                 signature=dbus.Signature('a{sv}'))
@@ -69,12 +69,13 @@ class RepoTest(unittest.TestCase):
     def test_list_repos_spec(self):
         # get list of specified repositories
         self.assertEqual(
-            self.iface_repo.list({"patterns": ['repo1']}),
+            self.iface_repo.list(
+                    {"repo_attrs": ["name", "enabled"], "patterns": ['rpm-repo1']}),
             dbus.Array([
                 dbus.Dictionary({
                     dbus.String('name'): dbus.String('Repository 1', variant_level=1),
                     dbus.String('enabled'): dbus.Boolean(True, variant_level=1),
-                    dbus.String('id'): dbus.String('repo1', variant_level=1)},
+                    dbus.String('id'): dbus.String('rpm-repo1', variant_level=1)},
                     signature=dbus.Signature('sv')),
                 ],
                 signature=dbus.Signature('a{sv}'))

--- a/test/dnfdaemon-server/test_repoquery.py
+++ b/test/dnfdaemon-server/test_repoquery.py
@@ -61,19 +61,19 @@ class RepoTest(unittest.TestCase):
             dbus.Array([
                 dbus.Dictionary({
                     dbus.String('full_nevra'): dbus.String('one-0:1-1.noarch', variant_level=1),
-                    dbus.String('repo'): dbus.String('repo1', variant_level=1)},
+                    dbus.String('repo'): dbus.String('rpm-repo1', variant_level=1)},
                     signature=dbus.Signature('sv')),
                 dbus.Dictionary({
                     dbus.String('full_nevra'): dbus.String('one-0:1-1.src', variant_level=1),
-                    dbus.String('repo'): dbus.String('repo1', variant_level=1)},
+                    dbus.String('repo'): dbus.String('rpm-repo1', variant_level=1)},
                     signature=dbus.Signature('sv')),
                 dbus.Dictionary({
                     dbus.String('full_nevra'): dbus.String('two-0:2-2.noarch', variant_level=1),
-                    dbus.String('repo'): dbus.String('repo2', variant_level=1)},
+                    dbus.String('repo'): dbus.String('rpm-repo2', variant_level=1)},
                     signature=dbus.Signature('sv')),
                 dbus.Dictionary({
                     dbus.String('full_nevra'): dbus.String('two-0:2-2.src', variant_level=1),
-                    dbus.String('repo'): dbus.String('repo2', variant_level=1)},
+                    dbus.String('repo'): dbus.String('rpm-repo2', variant_level=1)},
                     signature=dbus.Signature('sv')),
                 ],
                 signature=dbus.Signature('a{sv}'))
@@ -94,11 +94,11 @@ class RepoTest(unittest.TestCase):
             dbus.Array([
                 dbus.Dictionary({
                     dbus.String('full_nevra'): dbus.String('one-0:1-1.noarch', variant_level=1),
-                    dbus.String('repo'): dbus.String('repo1', variant_level=1)},
+                    dbus.String('repo'): dbus.String('rpm-repo1', variant_level=1)},
                     signature=dbus.Signature('sv')),
                 dbus.Dictionary({
                     dbus.String('full_nevra'): dbus.String('one-0:1-1.src', variant_level=1),
-                    dbus.String('repo'): dbus.String('repo1', variant_level=1)},
+                    dbus.String('repo'): dbus.String('rpm-repo1', variant_level=1)},
                     signature=dbus.Signature('sv')),
             ],
             signature=dbus.Signature('a{sv}'))

